src/Lykke.Service.Zcash.Api.AzureRepositories/Operations/OperationRepository.cs

```diff
@@ -0,0 +1,141 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using AzureStorage;
+using AzureStorage.Tables;
+using Common.Log;
+using Lykke.AzureStorage.Tables;
+using Lykke.Service.Zcash.Api.Core.Domain.Operations;
+using Lykke.SettingsReader;
+using Microsoft.WindowsAzure.Storage.Table;
+
+namespace Lykke.Service.Zcash.Api.AzureRepositories.Operations
+{
+    public class OperationRepository : IOperationRepository
+    {
+        private INoSQLTableStorage<OperationEntity> _operationStorage;
+        private INoSQLTableStorage<OperationItemEntity> _operationItemStorage;
+        private INoSQLTableStorage<IndexEntity> _indexStorage;
+        private static string GetOperationPartitionKey(Guid operationId) => operationId.ToString();
+        private static string GetOperationRowKey() => string.Empty;
+        private static string GetOperationItemPartitionKey(Guid operationId) => operationId.ToString();
+        private static string GetOperationItemRowKey() => string.Empty;
+        private static string GetIndexPartitionKey(string hash) => hash;
+        private static string GetIndexRowKey() => string.Empty;
+
+        public OperationRepository(IReloadingManager<string> connectionStringManager, ILog log)
+        {
+            _operationStorage = AzureTableStorage<OperationEntity>.Create(connectionStringManager, "ZcashOperations", log);
+            _operationItemStorage = AzureTableStorage<OperationItemEntity>.Create(connectionStringManager, "ZcashOperationItems", log);
+            _indexStorage = AzureTableStorage<IndexEntity>.Create(connectionStringManager, "ZcashOperationIndex", log);
+        }
+
+        public async Task<IOperation> UpsertAsync(Guid operationId, OperationType type, (string fromAddress, string toAddress, decimal amount)[] items,
+            decimal fee, bool subtractFee, string assetId)
+        {
+            var operationItemEntities = items.Select(item => new OperationItemEntity()
```

Comment:
operationEnumEntities will be enumerated twice. Convert to array after select.

Reply:
Fixed